### PR TITLE
Generate Method code action

### PR DESCRIPTION
**Pull request by BladeMF**
A code action for the "Generate method" refactoring. Some points (also noted in the code):
* It doesn't provide diagnostics (although the code is there, the tag is commented out from the container setup), because it doesn't check method calls for "undefined method". I wanted to push v1 out and mess around with (potentially) slow calls to Worse reflection for v2. So the whenever the code for `provideDiagnostics` is fitted with "check for undefined method", it will be good to go;
* As a result, if a method is defined and the code action is executed, nothing happens;
* It shows up on all method calls;
* Otherwise it's really nice to use.

The test I intentionally made a **unit** test rather than an **integration** test, because:
* Integration tests are really complicated to setup and I rather not do this without a reason;
* A single integration case for a `CodeActionProvider` and `DiagnosticsProvider` will prove the plumbing is working - I'd rather test each `CodeActionProvider` whether they provide the correct actions/diagnostics in the correct context and then test whether the `Command` is passing to and from the model and handling exceptions.

The only problem with the above is the very difficult mocking of the `ClientApi` class. We can create a `ClientApiSpy` class that collects all calls to `public function request(string $method, array $params): Promise` and `public function notification(string $method, array $params): void` methods or allows a callback to emulate responses.

### CODE REVIEW

File: lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                array_filter($diagnostics, function (Diagnostic $diag) use ($range) {
+                    return (
+                        $diag->range->start->line <= $range->start->line &&
+                        $diag->range->start->character <= $range->start->character &&
+                        $diag->range->end->line >= $range->end->line &&
+                        $diag->range->end->character >= $range->end->character
+                    );
+                })
+            );
+
```
Comment (BladeMF) [inline lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php:80]: The above probably needs to change to a code similar to what diagnostics does. I am generally fearful of performance similar to that of Import Names where it gets really slow. I wonder if we can cache the result of the diagnostics and use that to see whether we're on one? If the cache is not there we can always get them again.

File: lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            if (count($diagnostics) === 0) {
+                return [];
+            }
+
+            return [
+                CodeAction::fromArray([
+                    'title' =>  'Generate method (if not exists)',
```
Comment (BladeMF) [inline lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php:87]: This kind of goes with the fact that we're not checking the method calls.

File: lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                        ]
+                    )
+                ])
+            ];
+        });
+    }
+    /**
+     * @return array<Diagnostic>
+     */
+    private function getDiagnostics(TextDocumentItem $textDocument): array
```
Comment (BladeMF) [inline lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php:105]: Diagnostics are there, but they are not used in the LSP as they would underline every method call.

File: lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                    ),
+                ),
+                'Generate method',
+                DiagnosticSeverity::INFORMATION,
+                null,
+                'phpactor'
+            );
+        }
+
+        usort($diagnostics, function (Diagnostic $a, Diagnostic $b) {
```
Comment (BladeMF) [inline lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php:146]: If not sorted it becomes untestable because the order that `TolerantParser` returns the method calls is sometimes different from the order they occur in the source (purely as line/char).

File: lib/LanguageServerCodeTransform/LanguageServerCodeTransformExtension.php [hunk 5]
```diff
@@ -160,5 +176,15 @@ private function registerCodeActions(ContainerBuilder $container): void
...
             LanguageServerExtension::TAG_CODE_ACTION_PROVIDER => []
         ]);
+
+        $container->register(GenerateMethodProvider::class, function (Container $container) {
+            return new GenerateMethodProvider(
+                // $container->get(CodeTransformExtension::SERVICE_CLASS_GENERATORS),
+                $container->get('worse_reflection.tolerant_parser')
+            );
+        }, [
+            // LanguageServerExtension::TAG_DIAGNOSTICS_PROVIDER => [],
```
Comment (BladeMF) [inline lib/LanguageServerCodeTransform/LanguageServerCodeTransformExtension.php:184]: This is how the diagnostics are turned off.

File: lib/LanguageServerHover/Handler/HoverHandler.php [hunk 6]
```diff
@@ -73,16 +73,19 @@ public function hover(
...
-                PositionConverter::byteOffsetToPosition(
-                    ByteOffset::fromInt($symbolContext->symbol()->position()->start()),
-                    $document->__toString()
-                ),
-                PositionConverter::byteOffsetToPosition(
-                    ByteOffset::fromInt($symbolContext->symbol()->position()->end()),
-                    $document->__toString()
-                )
+                $position,
+                $position
```
Comment (BladeMF) [inline lib/LanguageServerHover/Handler/HoverHandler.php]: Not sure what to do about this one. It is really annoying but it probably belongs in its own PR.

File: tests/LanguageServerCodeTransform/Unit/LspCommand/GenerateMethodCommandTest.php [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        $sourceCode = SourceCode::fromStringAndPath(
+            $textDocumentItem->text,
+            $textDocumentUri->path()
+        );
+        $textEdits = new TextDocumentEdits(
+            $textDocumentUri,
+            new TextEdits(TextEdit::create(5, 1, 'test'))
+        );
+
+        $rpcClient = new class() implements RpcClient {
```
Comment (BladeMF) [inline tests/LanguageServerCodeTransform/Unit/LspCommand/GenerateMethodCommandTest.php]: By doing a mock `RpcClient` we can intercept calls to `notification` and `request` and then spy on them on mock responses.

File: lib/LanguageServerCodeTransform/LspCommand/GenerateMethodCommand.php [hunk 8]
```diff
@@ -61,6 +62,9 @@ public function __invoke(string $uri, int $offset): Promise
         } catch (MethodCallNotFound $error) {
             $this->clientApi->window()->showMessage()->warning($error->getMessage());
             return new Success(null);
+        } catch (CouldNotResolveNode $error) {
```
Comment (dantleech) [inline lib/LanguageServerCodeTransform/LspCommand/GenerateMethodCommand.php]: i think both of these extend `NotFound` ?

File: lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php (hunk 3 above)
Comment (dantleech) [reply lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php:105]: Why not detect if the method exists or not?

File: lib/LanguageServerHover/Handler/HoverHandler.php (hunk 6 above)
Comment (dantleech) [reply lib/LanguageServerHover/Handler/HoverHandler.php]: yes if not related please remove

File: tests/LanguageServerCodeTransform/Unit/LspCommand/GenerateMethodCommandTest.php (hunk 7 above)
Comment (dantleech) [reply tests/LanguageServerCodeTransform/Unit/LspCommand/GenerateMethodCommandTest.php]: you can use the test language server builder for this I think

Comment (dantleech) [reply tests/LanguageServerCodeTransform/Unit/LspCommand/GenerateMethodCommandTest.php]: https://github.com/phpactor/language-server-phpactor-extensions/blob/adf9f1857732cb0e1365d38f7614313ccf0f8f7a/tests/LanguageServerCodeTransform/Unit/CodeAction/CreateClassProviderTest.php#L65

the transmitter has all the messages that were transmitted from the LS

File: tests/LanguageServerCodeTransform/Unit/LspCommand/GenerateMethodCommandTest.php [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+            }
+        };
+        $apiClient = new ClientApi($rpcClient);
+
+        $workspace = $this->prophesize(Workspace::class);
+        // @phpstan-ignore-next-line
+        $workspace->get($uri)
+            ->willReturn($textDocumentItem);
+
+        $generateMethod = $this->prophesize(GenerateMethod::class);
```
Comment (dantleech) [inline tests/LanguageServerCodeTransform/Unit/LspCommand/GenerateMethodCommandTest.php:51]: if using the test language server you don't need to mock the workspace:

```
        $languageServer->textDocument()->open($url, $content)
```

File: lib/LanguageServerCodeTransform/LspCommand/GenerateMethodCommand.php (hunk 8 above)
Comment (BladeMF) [reply lib/LanguageServerCodeTransform/LspCommand/GenerateMethodCommand.php]: Indeed you are correct.

File: lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php (hunk 3 above)
Comment (BladeMF) [reply lib/LanguageServerCodeTransform/CodeAction/GenerateMethodProvider.php:105]: a) Was worried about performance
b) Wanted to get the other part running
c) Didn't know what was the best way to do that.

If you can help with c), that'll be great.

File: tests/LanguageServerCodeTransform/Unit/LspCommand/GenerateMethodCommandTest.php (hunk 7 above)
Comment (BladeMF) [reply tests/LanguageServerCodeTransform/Unit/LspCommand/GenerateMethodCommandTest.php]: Did it using `TestRpcClient`.

**Comment by BladeMF** [comment]
Not sure why 7.3 CS fixer fails...

**Comment by dantleech** [comment]
restarted the build, probably temporary failure. you should be able to use the LS server tester more than you have, but it's already much better by using RpcTestClient.

**Comment by dantleech** [comment]
:boom: I will see if I can iterate on this for diagnostics and the tests.

File: tests/LanguageServerCodeTransform/Unit/CodeAction/GenerateMethodProviderTest.php [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+use Phpactor\LanguageServerProtocol\DiagnosticSeverity;
+use Phpactor\LanguageServerProtocol\Position;
+use Phpactor\LanguageServerProtocol\Range;
+use Phpactor\LanguageServerProtocol\TextDocumentItem;
+use Phpactor\TextDocument\ByteOffsetRange;
+use function Amp\Promise\wait;
+
+class GenerateMethodProviderTest extends TestCase
+{
+    /** @dataProvider provideDiagnosticsTestData */
```
Comment (dantleech) [inline tests/LanguageServerCodeTransform/Unit/CodeAction/GenerateMethodProviderTest.php:24]: for future reference this should be:
```
/**
 * @dataProvider blah
 * /
 */
```

for consistence.

File: tests/LanguageServerCodeTransform/Unit/CodeAction/GenerateMethodProviderTest.php [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+                        GenerateMethodCommand::NAME,
+                        [
+                            $uri,
+                            $result->offset('selection')->toInt()
+                        ]
+                    )
+                ])
+            ];
+        } else {
+            $codeActions = [];
```
Comment (dantleech) [inline tests/LanguageServerCodeTransform/Unit/CodeAction/GenerateMethodProviderTest.php:130]: please try and avoid `else` statements in the future. `if (!$shouldSucceed) { self::assertEmpty($blah) }`

Comment (dantleech) [inline tests/LanguageServerCodeTransform/Unit/CodeAction/GenerateMethodProviderTest.php:130]: You can avoid this IF/ELSE by delegating `$codeActions = $this->createCodeActions($shouldSucceed)` for example and returning early.

File: tests/LanguageServerCodeTransform/Unit/LspCommand/GenerateMethodCommandTest.php [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        self::assertNotNull($showMessage);
+        self::assertEquals([
+            'type' => 2,
+            'message' => $exception->getMessage()
+        ], $showMessage->params);
+    }
+
+    public function provideExceptions(): array
+    {
+        return [
```
Comment (dantleech) [inline tests/LanguageServerCodeTransform/Unit/LspCommand/GenerateMethodCommandTest.php:125]: to be consistent ould be better in future to use `yield`

File: tests/LanguageServerCodeTransform/Unit/CodeAction/GenerateMethodProviderTest.php (hunk 10 above)
Comment (BladeMF) [reply tests/LanguageServerCodeTransform/Unit/CodeAction/GenerateMethodProviderTest.php:24]: I keep forgetting you want them multilined, sorry.

Comment (dantleech) [reply tests/LanguageServerCodeTransform/Unit/CodeAction/GenerateMethodProviderTest.php:24]: it's just consistency, I should update the CS fixer to catch it.
